### PR TITLE
PER-8985 feat: Playwright drop-in baseline seeding — build attributes + exec orchestration

### PR DESCRIPTION
**Pull request by Shivanshu-07**
## What

CLI-side support for the `@percy/playwright` toHaveScreenshot drop-in's baseline flow:

- **@percy/client** — `createBuild` accepts allow-listed drop-in build sources via `PERCY_BUILD_SOURCE` (`playwright-dropin` / `playwright-dropin-baseline`) plus `dropin-baseline-candidate` / `dropin-baseline-setup` attributes (env or explicit options). The API keys baseline auto-approval on these.
- **@percy/cli-exec** — generic baseline-provider discovery: any installed SDK package can declare `"@percy/cli": { "baselineProvider": "..." }` (no framework code enters the CLI, same scaling model as command discovery). On an EMPTY project, `percy exec` uploads the committed baseline screenshots as an auto-approved build #1 before the user's command runs (the suite never executes for the baseline), then starts the head build. Established projects get a notice pointing at the SDK's explicit `playwright:setup-baseline` command — the API's no-build-persisted sentinel guarantees no stray builds. Seeding dispatches by project type: web → rendered web snapshots (root DOM + image resource, the `percy upload` shape); app → comparison ingest (tag + tile, no render).
- **@percy/core** — the server-decided build `source` is kept on `percy.build` so SDKs can key on it through the healthcheck.

## Depends on

- percy/percy-playwright#640 (the SDK provider + command)
- percy/percy-api#6178 + companion API PR (server-side decision + auto-approval)

## Testing

Client + exec suites green incl. 11 new seeding specs; E2E verified against a local percy-api and on staging (empty web/app seed → head diff flow, established-project sentinel, explicit setup command).

## Deploy ordering

percy-api (percy/percy-api#6527) must ship first. Against an API that predates the `dropin-baseline-candidate` attribute, the compat guard abandons the seed build unused — on an *established* project that leaves one stray unfinalized build per run until the API side is live (they expire server-side, but numbering inflates).

## Automate project support (bf3b5b2e)

Automate tokens now flow through the drop-in gate: the provider walk still runs (so the head build carries the `playwright-dropin` source tag for API telemetry), but `maybeSeedBaseline` skips before touching discovery — Automate captures happen on the remote BrowserStack browser, so committed local screenshots can never pair with them; the first run establishes the baseline. Companion changes: percy/percy-playwright#640 (dispatch), percy/percy-api#6630 (server-side guards).
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Shivanshu-07** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2335  •  **Head:** 0a38a90  •  **Reviewers:** fallback inline checklist (deep-review agent)

## Summary
CLI-side drop-in baseline seeding: generic `baselineProvider` package discovery, seed-build creation/upload (`percy exec` seeds an empty project's build #1 from committed screenshots, then waits for it to finish before the head build starts), new allow-listed build source + candidate/setup attributes in @percy/client, plus two @percy/core coverage fixes inherited from master.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | |
| High | Security | Authentication/authorization checks present | N/A | token handled by @percy/client env |
| High | Security | Input validation and sanitization | Pass | sources allow-listed (DROPIN_BUILD_SOURCES); env flags strict `=== 'true'` |
| High | Security | No IDOR — resource ownership validated | N/A | |
| High | Security | No SQL injection (parameterized queries) | N/A | |
| High | Correctness | Logic is correct, handles edge cases | Pass | 2 High findings fixed in 0a38a90 lineage: discovery walk now degrades to [] on fs errors (a throw previously escaped to exec's catch and disabled Percy for the run); seeding is bounded to build #1 so a pre-candidate API can no longer accrete stray baseline builds on established projects |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | never-throw seeding posture is by design; walk failures logged at debug |
| High | Correctness | No race conditions or concurrency issues | Pass | seed-wait closes the head-baseline race; parallel shards skip seeding |
| Medium | Testing | New code has corresponding tests | Pass | walk-error, containment-escape, homedir-bound, non-first-build, falsy-entry cases added |
| Medium | Testing | Error paths and edge cases tested | Pass | |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | all package suites + 100% coverage gates green |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | walk now stops at the nearest node_modules (was: every ancestor to /) |
| Medium | Performance | Long-running tasks use background jobs | N/A | |
| Medium | Quality | Follows existing codebase patterns | Pass | walk mirrors @percy/cli findModulePackages semantics exactly |
| Medium | Quality | Changes are focused (single concern) | Pass | the two @percy/core fixes unblock master's own red Test workflow |
| Low | Quality | Meaningful names, no dead code | Pass | |
| Low | Quality | Comments explain why, not what | Pass | |
| Low | Quality | No unnecessary dependencies added | Pass | |

## Findings

All review findings were fixed on this PR before this report:
- **High** `packages/cli-exec/src/baseline.js`: the node_modules walk could throw (EACCES / node_modules-as-file) out of `findBaselineProvider` into exec's outer catch — silently disabling Percy for the whole run. Walk now try/caught → `[]`, matching the command-discovery pattern.
- **High** `packages/cli-exec/src/baseline.js`: against an API that predates the candidate attribute, every run would seed a stray "baseline" build into established projects. Seeding now requires `build-number === 1`.
- **Medium** `waitForSeedBuild` uses `client.getBuild` — verified working with standard write project tokens against percy-api (staging E2E), so the read-token doc caveat does not bite; kept as-is.
- **Low**: walk bounded at the home directory and stops at the nearest node_modules (narrower import surface); provider modules confined to their declaring package; falsy provider entries filtered instead of terminating an upload worker; containment paths wrapped in the sanctioned sanitizer idiom (semgrep).

---

**Verdict: PASS**

File: packages/cli-exec/src/baseline.js [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+// processed, and an unapproved seed means the whole first run shows as new instead of diffing.
+// The timeout matches the pipeline latency budget (~99% of builds finish under 5 minutes) —
+// a seed of committed screenshots still renders server-side, so first runs can hold for minutes.
+export async function waitForSeedBuild(client, buildId, { log, timeout = 600000, interval = 5000 }) {
+  let deadline = Date.now() + timeout;
+  let state = 'pending';
+  let polls = 0;
+
+  for (;;) {
+    ({ state } = (await client.getBuild(buildId)).data.attributes);
```
Comment (aryanku-dev) [inline packages/cli-exec/src/baseline.js:138]: `client.getBuild`'s own doc comment in `@percy/client` says it requires a **read access** token, but most `percy exec` users run with a plain write-only project token. If `GET /builds/:id` 403s here, the first poll throws, `maybeSeedBaseline` swallows it, and every first run warns `did not finish processing in time (state: undefined)` — then starts the head build immediately, racing the unapproved seed. That silently defeats the seed-before-head ordering this feature depends on.

Can we verify this loop works on staging with a standard write-only token (not a full-access one)? If it 403s, the API needs to allow build-state reads for the token that created the build.

File: packages/cli-exec/src/baseline.js [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+    let buildId = res.data.id;
+    log.info(`New Percy project with ${baselines.length} committed baseline snapshot(s) ` +
+      'detected — establishing your baseline (build #1) before running tests');
+
+    let seeded = await uploadBaselines(percy.client, buildId, baselines, {
+      log, projectType: percy.projectType
+    });
+
+    await percy.client.finalizeBuild(buildId);
```
Comment (aryanku-dev) [inline packages/cli-exec/src/baseline.js:217]: If every upload fails transiently (e.g. a network blip), `seeded` is 0 but we still finalize — creating an **empty, auto-approved build #1**. The project is now "established" with an empty baseline: the next run diffs against nothing *and* seeding never retries (build-number is no longer 1), so a one-time hiccup permanently burns the auto-seed path.

Suggest: when `seeded === 0`, skip the finalize (or at least don't log "Baseline established") and point at `percy playwright:setup-baseline` instead. Partial seeding is fine; zero is not.

File: packages/cli-exec/src/baseline.js [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+      return false;
+    }
+    // A provider yielding sparse/null entries must not wedge the upload workers.
+    baselines = baselines.filter(Boolean);
+    if (!baselines.length) return false;
+
+    // Ask the server. An explicit baseline source on an ESTABLISHED project returns the
+    // baseline-skipped sentinel (no build persisted); on an empty project it creates build #1 as
+    // the baseline. First-ness is decided by the API from the project token — never locally.
+    let res = await percy.client.createBuild({
```
Comment (aryanku-dev) [inline packages/cli-exec/src/baseline.js:182]: Once a project has builds, every subsequent `percy exec` with committed baselines still pays this full cost on **every run, forever**: the provider's `discoverBaselines()` filesystem walk, plus a doomed `createBuild` POST the API answers with the sentinel, plus two info log lines ("already has builds — skipping" / "run npx percy playwright:setup-baseline"). That's permanent log noise and an extra API round-trip per run for the lifetime of the project.

Consider gating the notice so it only shows once, or having the head build response carry an "established" hint so later runs can skip the sentinel round-trip entirely.

File: packages/cli-exec/src/baseline.js (hunk 1 above)
Comment (Shivanshu-07) [reply packages/cli-exec/src/baseline.js:138]: Verified empirically on staging with standard write-only project tokens (`web_…`/`app_…`, not full-access): the wait loop polled through `pending → processing → finished` and the seed auto-approved — e.g. seed builds 316760 (web) and 316793 (app) on staging org c841aa36, both driven end-to-end by `percy exec` with the plain project token. percy-api allows token-auth build-state reads for the token's own project, so the client-side doc comment is conservative rather than binding here. Also hardened the failure message in e2b4399b: a failed poll now reports `state: unknown` (with the underlying error at debug) instead of `undefined`.

File: packages/cli-exec/src/baseline.js (hunk 2 above)
Comment (Shivanshu-07) [reply packages/cli-exec/src/baseline.js:217]: Great catch — real bug. Fixed in e2b4399b: `seeded === 0` now skips the finalize entirely (the unfinalized build expires server-side, so nothing establishes), warns `baseline not established`, and points at `percy playwright:setup-baseline` / re-running once connectivity recovers. Partial seeding still finalizes as before. Unit test added pinning no-finalize + the warning.

File: packages/cli-exec/src/baseline.js (hunk 3 above)
Comment (Shivanshu-07) [reply packages/cli-exec/src/baseline.js:182]: Fair point on the noise — the two info lines are now one (e2b4399b). On skipping the round-trip entirely: the sentinel POST is answered instantly with nothing persisted, so the real per-run cost is discovery's config resolution, and reordering the sentinel BEFORE discovery would mean creating a candidate build before knowing the repo has baselines — an empty project with no committed PNGs would then accrete a stray pending build every run, which is worse. A cross-run skip needs client-side persistent state (no precedent in the CLI) or a project-state read before build creation; happy to file a follow-up if the one-line-per-run notice still feels too loud.

**Comment by Shivanshu-07** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2335  •  **Head:** da65d31  •  **Reviewers:** fallback inline checklist (deep-review agent, re-review)

## Summary
Re-review after the master merge, reviewer-feedback fixes (zero-upload seed no longer finalizes), and comment cleanup. CLI-side drop-in baseline seeding: provider discovery, seed build #1 from committed screenshots with a bounded wait before the head build, allow-listed build sources, and the `playwright:setup-baseline` command.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | |
| High | Security | Authentication/authorization checks present | N/A | token handled by @percy/client env |
| High | Security | Input validation and sanitization | Pass | env flags strict `=== 'true'`; sources allow-listed; rogue provider buildSource cannot inject |
| High | Security | No IDOR — resource ownership validated | N/A | |
| High | Security | No SQL injection (parameterized queries) | N/A | |
| High | Correctness | Logic is correct, handles edge cases | Pass | zero-seed path verified (no finalize, no wait, never throws); walk narrowed to percy-cli-* matching findModulePackages |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | never-throw seeding posture holds on every path |
| High | Correctness | No race conditions or concurrency issues | Pass | seed-wait now tick-responsive to Ctrl-C via yieldTo (fixed in da65d31); worker queue race-free |
| Medium | Testing | New code has corresponding tests | Pass | zero-seed, established sentinel, pre-candidate guard, walk bounds, env fallbacks (setup variant added in da65d31) |
| Medium | Testing | Error paths and edge cases tested | Pass | |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | all package suites + 100% coverage gates green (Windows core flake was a degraded runner on a master-owned spec) |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | walk stops at nearest node_modules, home-dir bound |
| Medium | Performance | Long-running tasks use background jobs | N/A | |
| Medium | Quality | Follows existing codebase patterns | Pass | HTML shape mirrors percy upload; walk mirrors command discovery |
| Medium | Quality | Changes are focused (single concern) | Pass | deploy-ordering note (api PR first) added to the PR description |
| Low | Quality | Meaningful names, no dead code | Pass | |
| Low | Quality | Comments explain why, not what | Pass | zero-seed recovery text corrected (setup-baseline, not re-run) |
| Low | Quality | No unnecessary dependencies added | Pass | |

## Findings

Re-review findings (2 Medium, 3 Low), all fixed on this PR in da65d31:
- **Medium** `packages/cli-exec/src/exec.js`: the up-to-10-minute seed wait was the one stretch the abort signal couldn't unwind — a first Ctrl-C appeared ignored. Now driven through `yieldTo`, keeping the runner tick-responsive.
- **Medium** `packages/cli-exec/src/baseline.js`: the zero-seed warning promised a plain re-run recovers, but a stray unfinalized candidate build can bump the build number past the pre-candidate compat guard — recovery text now points at `percy playwright:setup-baseline`, and the guard's comment states its actual role.
- **Low**: unscoped walk collection narrowed from `percy-*` to `percy-cli-*` (matches `findModulePackages`, smaller auto-import surface); deploy-ordering requirement documented in the PR body; env-fallback test added for the setup attribute.

Earlier human-review threads (aryanku-dev): all resolved — write-token wait verified empirically on staging; zero-upload finalize fixed; established-project notice reduced to one line.

---

**Verdict: PASS**

**Comment by Shivanshu-07** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2335  •  **Head:** bf3b5b2  •  **Reviewers:** fallback inline checklist

## Summary
Incremental review of the automate-support push (`26875421` + master merge `bf3b5b2e`; everything up to `da65d313` was covered by the previous PASS): the exec seeding gate now includes automate tokens so the head build carries the `playwright-dropin` source tag, while `maybeSeedBaseline` skips before touching discovery — remote Automate captures can never pair with committed local screenshots.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Test tokens are placeholder strings (`auto_PERCY_TOKEN`, `ss_PERCY_TOKEN`) |
| High | Security | Authentication/authorization checks present | N/A | No auth surface touched; server-side enforcement lives in percy/percy-api#6630 |
| High | Security | Input validation and sanitization | Pass | No new input paths; the guard only narrows behavior by project type |
| High | Security | No IDOR — resource ownership validated | N/A | No resource access changes |
| High | Security | No SQL injection (parameterized queries) | N/A | No queries |
| High | Correctness | Logic is correct, handles edge cases | Pass | Automate skip runs FIRST in `maybeSeedBaseline` (before the parallel-shard check and provider discovery), inside the never-throws try; the exec gate tags `PERCY_BUILD_SOURCE` before build creation. Verified live on staging: build 317035 `source=playwright-dropin`, seeding-skip debug line present |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Skip path returns `false` like every other decline branch; no new throw paths |
| High | Correctness | No race conditions or concurrency issues | Pass | Env-var tag is set once before `percy.yield.start()`, same ordering as the existing web/app path |
| Medium | Testing | New code has corresponding tests | Pass | 2 new specs: exec tags-but-never-seeds for automate (asserts the debug line + env tag), and `maybeSeedBaseline` skips without invoking `discoverBaselines` |
| Medium | Testing | Error paths and edge cases tested | Pass | Generic (`ss_`) token still skips discovery entirely — re-pinned after the gate change inverted the old automate expectation |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | CI green 46/46 on head; local cli-exec run's single failure is the pre-existing SIGTERM-interrupt timing flake (fails identically on the pristine tree) |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Automate now runs the provider walk (bounded, nearest-node_modules) but skips discovery/upload |
| Medium | Performance | Long-running tasks use background jobs | N/A | The ≤10min seed wait is never entered for automate |
| Medium | Quality | Follows existing codebase patterns | Pass | Matches the existing decline-branch style (`log.debug` + `return false`) and the gate's comment conventions |
| Medium | Quality | Changes are focused (single concern) | Pass | +50/−3 across 4 files; the merge commit only syncs master (1.32.5) |
| Low | Quality | Meaningful names, no dead code | Pass | No new symbols |
| Low | Quality | Comments explain why, not what | Pass | Comments carry the pairing rationale, not mechanics |
| Low | Quality | No unnecessary dependencies added | Pass | None |

## Findings

No findings — no Fail rows.

Note: the head commit is a base-branch merge, but the same push introduced the unreviewed `26875421`, so the merge fast-path was deliberately not taken and the novel diff was reviewed in full.

---

**Verdict: PASS**

**Review by aryanku-dev — APPROVED** [review]
